Src/Support/Google.Apis.Auth.AspNetCore/GoogleAuthProvider.cs

```diff
@@ -0,0 +1,118 @@
+            var expiresUtc = auth.Properties.ExpiresUtc;
+            if (string.IsNullOrEmpty(accessToken) || string.IsNullOrEmpty(refreshToken) || expiresUtc == null)
+            {
+                throw new InvalidOperationException("Invalid auth. access_token, refresh_token, and expires_utc must all be present.");
+            }
+            if (expiresUtc.Value - (accessTokenRefreshWindow ?? s_accessTokenRefreshWindow) < _clock.UtcNow)
+            {
+                // Refresh required. This has to be done inline here (it can't be done in the background)
+                // because the request auth properties need to be updated with the result.
+                var options = _options.Get(_scheme);
+                var oidcConfig = await options.ConfigurationManager.GetConfigurationAsync(cancellationToken);
+                var refreshContent = new FormUrlEncodedContent(new Dictionary<string, string>
+                {
+                    { "client_id", options.ClientId },
+                    { "client_secret", options.ClientSecret },
+                    { "grant_type", "refresh_token" },
+                    { "refresh_token", auth.Properties.GetTokenValue("refresh_token") }
+                });
+                try
+                {
+                    var refreshResponse = await options.Backchannel.PostAsync(oidcConfig.TokenEndpoint, refreshContent, cancellationToken);
+                    var payload = JObject.Parse(await refreshResponse.Content.ReadAsStringAsync());
+                    var refreshedAccessToken = payload.Value<string>("access_token");
+                    var refreshedRefreshToken = payload.Value<string>("refresh_token");
+                    var refreshedExpiresIn = payload.Value<string>("expires_in");
+                    auth.Properties.UpdateTokenValue(OpenIdConnectParameterNames.AccessToken, refreshedAccessToken);
+                    if (!string.IsNullOrEmpty(refreshedRefreshToken))
+                    {
+                        auth.Properties.UpdateTokenValue(OpenIdConnectParameterNames.RefreshToken, refreshedRefreshToken);
+                    }
+                    auth.Properties.IssuedUtc = _clock.UtcNow;
+                    if (int.TryParse(refreshedExpiresIn, out int expiresInSeconds))
+                    {
+                        auth.Properties.ExpiresUtc = _clock.UtcNow + TimeSpan.FromSeconds(expiresInSeconds);
+                    }
+                }
+                catch (Exception e)
+                {
+                    throw new InvalidOperationException("Failed to refresh access_token.", e);
+                }
+                // Sign-in, to store the refreshed auth tokens (often stored into a cookie).
+                await httpContext.SignInAsync(options.SignInScheme, auth.Principal, auth.Properties);
+            }
+            // Return a short-term, non-refreshable credential.
+            return GoogleCredential.FromAccessToken(accessToken);
```

Comment:
If we've just refreshed the access token, why do we still return the old one?

Reply:
This is a bug.
I've added a test to catch this, and fixed it.
Thanks for spotting :)